src/input/joystick.c

```diff
@@ -0,0 +1,714 @@
+{
+    memset(joystick->button, 0, sizeof(joystick_button) * JOYSTICK_MAX_BUTTONS);
+}
+
+static void joystick_reset_all_axis_states(joystick_info *joystick)
+{
+    memset(joystick->axis, 0, sizeof(int) * JOYSTICK_MAX_AXIS);
+}
+
+static void joystick_reset_all_hat_states(joystick_info *joystick)
+{
+    memset(joystick->hat, 0, sizeof(joystick_hat) * JOYSTICK_MAX_HATS);
+}
+
+static void joystick_reset_all_trackball_states(joystick_info *joystick)
+{
+    memset(joystick->trackball, 0, sizeof(joystick_trackball) * JOYSTICK_MAX_TRACKBALLS);
+}
+
+static void joystick_reset_state(joystick_info *joystick)
+{
+    joystick_reset_all_axis_states(joystick);
+    joystick_reset_all_button_states(joystick);
+    joystick_reset_all_trackball_states(joystick);
+    joystick_reset_all_hat_states(joystick);
+}
+
+static joystick_info *get_free_joystick(void)
+{
+    for (int i = 0; i < JOYSTICK_MAX_CONTROLLERS; ++i) {
+        if (!data.joystick[i].connected) {
+            return &data.joystick[i];
+        }
+    }
+    return 0;
+}
+
+static joystick_info *get_joystick_from_id(int joystick_id)
+{
+    for (int i = 0; i < JOYSTICK_MAX_CONTROLLERS; ++i) {
+        if (data.joystick[i].id == joystick_id && data.joystick[i].connected) {
+            return &data.joystick[i];
+        }
+    }
+    return 0;
+}
+
+int joystick_add(int joystick_id, joystick_model *model)
+{
+    joystick_info *joystick = get_free_joystick();
+    if (!joystick) {
+        return 0;
+    }
+    joystick->id = joystick_id;
+    joystick->model = model;
+    joystick->connected = 1;
+    model->connected_joysticks++;
+    data.connected_joysticks++;
+    return 1;
+}
+
+int joystick_is_listened(int joystick_id)
+{
+    return get_joystick_from_id(joystick_id) != 0;
+}
+
+int joystick_remove(int joystick_id)
+{
+    joystick_info *joystick = get_joystick_from_id(joystick_id);
+    if (!joystick) {
+        return 0;
+    }
+    joystick->connected = 0;
+    joystick->model->connected_joysticks--;
+    joystick->model = 0;
+    joystick_reset_state(joystick);
+    data.connected_joysticks--;
+    return 1;
+}
+
+void joystick_update_element(int joystick_id, joystick_element element, int element_id, int value1, int value2)
+{
+    joystick_info *joystick = get_joystick_from_id(joystick_id);
+    if (!joystick) {
+        return;
+    }
+    switch (element) {
+        case JOYSTICK_ELEMENT_BUTTON:
+            joystick->button[element_id] = value1;
+            break;
+        case JOYSTICK_ELEMENT_HAT:
+            update_hat(&joystick->hat[element_id], value1);
+            break;
+        case JOYSTICK_ELEMENT_AXIS:
+            joystick->axis[element_id] = (abs(value1) > 1300) ? value1 : 0;
+            break;
+        case JOYSTICK_ELEMENT_TRACKBALL:
+            update_trackball(&joystick->trackball[element_id], value1, value2);
+            break;
+        default:
+            log_info("Trying to update wrong joystick element", 0, element);
```

Comment:
Also end this case with a `break;`

Reply:
Sure, but why is there a need to end the last case with a `break;`?

Reply:
For consistency's sake and defensive programming. See also: [here](https://softwareengineering.stackexchange.com/a/201786) and [here](https://stackoverflow.com/a/26139061).

Reply:
Makes sense! Thanks for the explanation